File: polymorphism/polymorphism_overloading.js
//Polymorphism reduces coupling in our application 

//Allows to create more compact code. 

//A language that supports polymorphism supports: 
//Overloading 
//Parametric Polymorphism 
//Inclusion polymorphism 

function countItems(x) { 
    return x.toString().length;
}

console.log(countItems(1000))
console.log(countItems("hola mundo"))

function sum(x = 0, y = 0, z = 0) { 
    return x + y + z 
}

console.log(sum(10,20))
console.log(sum(10,20,30))

function Stack() { 
    this.items = [] 

    this.push = function(item) { 
        this.items.push(item)
    }

}

const stack = new Stack()
stack.push('asdasd')
const stack2 = new Stack()
stack2.push(1000)

console.log(stack)
console.log(stack2)


//SUBTLE POLYMORPHISM 

class Person { 
    constructor(name, lastname) {
        this.name = name; 
        this.lastname = lastname;


    }

    
}

class Programmer extends Person { 
    constructor(language) { 
        super() 
        this.language = language
    }
}

const john = new Person('john', 'ray');
const ryan = new Programmer('javascript','ryan', 'ray');

console.log(john) 
console.log(ryan)

function writeFullName(p) { 
    console.log(p.name + " " + p.lastname)
}

writeFullName(john) 
writeFullName(ryan)

File: polymorphism/polymorphism_overloading.cs
using system; 

class Program { 
    public static void Main (string[] args) { 
        Console.Writeline("Hello World!")

        Program program = new Program()
        Console.Writeline(program.CountItems(30000));
        Console.Writeline(program.CountItems("Hola mundo"));

        Console.Writeline(program.Sum(10,20));
        Console.Writeline(program.Sum(10,20,30));

      var stack = new Stack<String>(); 
      var stack2 = new Stack<int>();

      var john = new Person(); 
      john.Name = "John"; 
      john.Lastname = "Ray";

      program.WriteFullName(john);

      var ryan = new Programmer
      ryan.Name = "Ryan"; 
      ryan.Lastname = "Ray"; 
      ryan.Language = "C#"

        program.WriteFullName(ryan); 
    }

    //Overloading 

    public int CountItems(int x) { 
        return x.ToString().Length;
    }

    public int CountItems(string x) { 
        return x.Length;
    }

    public int Sum(int x, int y) { 
        return Sum(x, y, 0); 
    }
    
    public int Sum(int x, int y, int z) { 
        return x + y + z; 
    }

    public void WriteFullName(Person p) { 
        Console.Writeline(p.Name + " " + p.Lastname);
    } 

}

//GENERICS 

public class Stack<T> { 
    private T[] items; 
    public void Push() { // ACCESS MODIFIERS 

    }  
}

//SUBTLE POLYMORPHISM 

public class Person { 
    public string Name {get; set; }
    public string Lastname {get; set;}
}

public class Programmer: Person { 
    public string Language {get; set;}
}